Demos (#13)
eProsima-Private/AML-Tools: docs/rst/demos/queens_embedding.rst

.. include:: /rst/exports/roles.include

.. _demos_queens_embedding:

Create an AML Embedding
#######################

In this section we describe step by step how to create an embedding.
We illustrate the process by creating an embedding for the N-Queens Completion Problem.

> The amldl package provides all the functions needed to build an embedding.

Reviewer: ```suggestion
The ``amldl`` package provides all the functions needed to build an embedding.
```

For the N-Queens Completion Problem we use the following functions:

* **Descriptor**: object that represents the embedding. It works as a Python context manager. AMLDL functions only work within such context.
* **ADD**: commit a duple to the embedding.
* **APP**: append an element to the end of a vector.
* **C**: create a constants.
* **V**: create an empty vector.
* **CV**: create a vector of constants.
* **INC**: create positive duple (inclusion).
* **EXC**: create negative duple (exclusion).
* **F**: find (by name or by position in a vector).
* **M**: merge constants.
* **R**: remove constant.
* **S**: create set from vector.
* **T**: tensorial operations, it allows for compact for-like loops.

.. image:: board_queens.png
   :width: 250
   :alt: One of the solution to the 8-Queens completion problem
   :align: center

Structure
*********

> Embedding contain two main sections.

Reviewer: ```suggestion
Embedding contain two main sections: the definition of constants and the definition of duples. 
```

Author: Embedding**S** contain [...]

> Embedding contain two main sections.

Reviewer: ```suggestion
Embeddings contain two main sections.
```

> First, the definition of constants.

Reviewer: ```suggestion
Let's start with the definition of constants.
```

The second part of the embedding is the definition of duples.
Duples allow us to establish relations between those complex idea that we can build with constants.
For instance, we can build a board as described above and also the concept of being a solution, with those two ideas we can build a positive relation that certain arrangement of pieces and empty cells conform a solution. Likewise, we can build the concept of a rule, for instance constructing a duple for two queens on the same row not being a correct solution.

Additionally, we can add some preprocessing before returning the embedding.
In the example below, we return just the duples and constants definitions, without returning the whole embedding object, the Descriptor.

.. code-block:: python

    # Algebraic AI - Copyright - All Rights Reserved

    from amldl import ADD, APP, C, CV, Descriptor, EXC, F, INC, M, R, T, V, S

    boardDim = 8

    def embedding(args):

        def ix(x, y):
            return x + y * boardDim

        with Descriptor() as theEmbedding:

            # -- Define constants

            # -- Define duples

            # -- Prepare output

            return

> Since vectors are one-dimensional in AMLDL, we make use of an auxiliary function to transform coordinates.

Reviewer: ```suggestion
Since vectors are one-dimensional in AML-DL, we make use of an auxiliary function to transform coordinates.
```

This is just used for readability.

Defining Constants
******************

> Constants are defined using the C command.

Reviewer: ```suggestion
Constants are defined using the ``C`` command.
```

Vectors are used to organise related constants.

> And the CV command serves as a shortcut to create families of constants in the form of vectors.

Reviewer: ```suggestion
And the ``CV`` command serves as a shortcut to create families of constants in the form of vectors.
```

    CV("Q", boardDim * boardDim)
    # Vector of Empty cell at (i,j) [Flatten 2D matrix]
    CV("E", boardDim * boardDim)
    # Vector of Rows. Row(i) indicates that there's at least one Q in row i.
    CV("R", boardDim)
    # Vector of Columns. Columns(j) indicates that there's at least one Q in column j.
    CV("C", boardDim)
    # A complete board, one that is a solution to the problem
    C("B")

    # Set with all possible options
    QE = M("Q", "E")

    # Create a list of vectors with all Q's or E's for a certain coordinate
    # E.g. Q(x=, ) = [Vx=0, Vx=1, ...]
    #   where Vx=0 = [Q(0,0), Q(0,1), Q(0,2), ...]
    V("Q(x=, )")
    V("Q( ,y=)")
    V("E(x=, )")
    V("E( ,y=)")
    for _ in range(0, boardDim):
        APP(F("Q(x=, )"), V())
        APP(F("Q( ,y=)"), V())
        APP(F("E(x=, )"), V())
        APP(F("E( ,y=)"), V())
    for x in range(0, boardDim):
        for y in range(0, boardDim):
            APP(F("Q(x=, )", x), F("Q", ix(x, y)))
            APP(F("Q( ,y=)", y), F("Q", ix(x, y)))
            APP(F("E(x=, )", x), F("E", ix(x, y)))
            APP(F("E( ,y=)", y), F("E", ix(x, y)))

Defining Duples
***************

> The INC and EXC commands are used to create duples.

Reviewer: ```suggestion
The ``INC`` and ``EXC`` commands are used to create duples.
```

> These duples then need to be added to the embedding using the ADD command.

Reviewer: ```suggestion
These duples then need to be added to the embedding using the ``ADD`` command.
```

                ADD(INC(M(*diags), M(Qxy, "B")))

    for x in range(0, boardDim):
        for y in range(0, boardDim):
            diags = []
            for xx in range(0, boardDim):
                for yy in range(0, boardDim):
                    if (x != xx) and (y != yy):
                        if (x - xx) == -(y - yy):
                            Exxyyy = F("E", ix(xx, yy))
                            diags.append(Exxyyy)
            Qxy = F("Q", ix(x, y))
            if diags:
                ADD(INC(M(*diags), M(Qxy, "B")))

    # Set some Queens as boundary conditions
    theEmbedding.REGION = 9
    if boardDim == 5:
        # Set a Queen at the center cell
        ADD(INC(F("Q", ix(2, 2)), "B"))

    if boardDim == 8:
        # Set two Queens at correct positions
        ADD(INC(M(F("Q", ix(1, 4)), F("Q", ix(3, 3))), "B"))

    # Both, empty and queen are not allowed in the solution board
    # Q(x,y) !< {E(x,y), B} for all (x,y)
    theEmbedding.REGION = 11
    ADD(EXC(M(T("Q", _i), T("E", _i)), "B"))

    # Require all columns and all rows with a queen
    # {R1, R2, ..., C1, C2...} < B
    theEmbedding.REGION = 12
    ADD(INC("R", "B"))
    ADD(INC("C", "B"))

> The T command can be complex to use and it can generally be expanded into a normal loop:

Reviewer: ```suggestion
The ``T`` command can be complex to use and it can generally be expanded into a normal loop:
```

.. code-block:: python

    EXC(T("R", _i), M(R("Q(x=, )", T("Q(x=, )", _i)), "E"))

    term_left = T("R", _i)
    term_right = M(R("Q(x=, )", T("Q(x=, )", _i)), "E")

> The term on the left sets a loop over elements in the R term (representing active Rows)

Reviewer: ```suggestion
The term on the left sets a loop over elements in the ``R`` term (representing active Rows)
```

> The term on the right combines the result of an R command with the term E (all Empty cells).
> The R command removes from the Q(x=,) vector, the term at a certain position pointed by the T operator.
> Since both T's use the same index, they move alongside (as in a one-dimensional loop).

Reviewer: ```suggestion
The term on the right combines the result of an ``R`` command with the term ``E`` (all Empty cells).
The ``R`` command removes from the ``Q(x=,)`` vector, the term at a certain position pointed by the ``T`` operator.
Since both ``T``'s use the same index, they move alongside (as in a one-dimensional loop).
```
